lib/WUI/http_lifetime.cpp

```diff
@@ -0,0 +1,196 @@
+#include "nhttp/server.h"
+#include "nhttp/common_selectors.h"
+#include "nhttp/headers.h"
+#include "nhttp/static_mem.h"
+#include "nhttp/gcode_upload.h"
+#include "wui.h"
+#include "wui_api.h"
+#include "wui_REST_api.h"
+
+#include <FreeRTOS.h>
+#include <semphr.h>
+#include <lwip/apps/fs.h>
+
+#include <cassert>
+#include <cstring>
+
+using std::nullopt;
+using std::optional;
+using std::string_view;
+using namespace nhttp;
+using namespace nhttp::handler;
+using namespace nhttp::handler::selectors;
+using nhttp::printer::GcodeUpload;
+
+namespace {
+
+SemaphoreHandle_t httpd_mutex = NULL;
+
+// TODO: Move to http_handler_default.cpp?
+// TODO: Eventually remove in favor of files in XFlash.
+class StaticFsFile final : public Selector {
+public:
+    virtual optional<ConnectionState> accept(const RequestParser &parser) const override {
+        char filename[MAX_URL_LEN + 1];
+
+        if (!parser.uri_filename(filename, sizeof(filename))) {
+            return nullopt;
+        }
+
+        if (strcmp(filename, "/") == 0) {
+            strncpy(filename, "/index.html", MAX_URL_LEN);
+        }
+
+        fs_file file;
+        if (fs_open(&file, filename) == ERR_OK) {
+#if LWIP_HTTPD_CUSTOM_FILES
+            // These are not static chunk of memory, not supported.
+            if (file.is_custom_file) {
+                fs_close(&file);
+                return nullopt;
+            }
+#endif
+            /*
+             * The data of non-custom fs files are embedded inside the program
+             * as C arrays. Therefore, we can get the data and close the file
+             * right away.
+             */
+            /*
+             * We guess the content type by a file extension. Looking inside
+             * and doing "file magic" would be better/more reliable, but don't
+             * want the code complexity of that.
+             */
+            static const char *extra_hdrs[] = { "Content-Encoding: gzip\r\n", nullptr };
```

Comment:
For now we only have gzipped static data, so the extra header seems to be fine. What happens once someone tries to return a file which is not gzipped?

Reply:
These specific „static files“ are for serving PrusaLink pages and they are all gzipped. We'll be getting rid of this soon (hopefully) in favor XFlash.

Anyway, if we ever want to also return some other (non-gzipped) „static files“, then we would create a different selector, or have a whitelist/blacklist for the urls that doesn't pass these extra headers :shrug:

Reply:
ok